bitcoin/src/address.rs

```diff
@@ -219,7 +219,12 @@ pub enum Payload {
     /// P2SH address.
     ScriptHash(ScriptHash),
     /// Segwit address.
-    WitnessProgram(WitnessProgram),
+    Segwit {
+        /// Witness version (version byte of the scriptPubkey).
+        version: WitnessVersion,
+        /// Witness program (as defined by BIP141).
+        program: WitnessProgram,
```

Comment:
So this no longer protects the invariant. I think a struct would be better. I guess calling it `Segwit` would be fine but I'm too tired to say for sure.

Reply:
I believe this is resolved but the original code that this talks about is so long ago I don't really remember it. We do have a struct variant now.

Reply:
I think Kix is right here. We can create the following invalid value for the Payload type.

```rust
let prog = WitnessProgram::new(WitnessVersion::V2, &[0x02; 31]).unwrap(); //31 bytes v2 progs are okay
let payload = Payload::Segwit {version: WitnessVersion::V0, program: prog}; // 31 bytes v0 are incorrect.
```

Reply:
I was happy to find that I did actually notice this when I coded it up, its in the patch description
```
    This change makes the code more semantically correct. Note that it does
    introduce a non-explicit assumption that the same version be passed to
    `WitnessProgram::new` as is used in the `Payload` - is this acceptable
    engineering?
```

After attempting to fix this I ended up creating a struct that was exactly the same as the original `WitnessProgram` but with a different name. I've elected to drop the second patch and just add the following comment to the `WitnessProgram`
```
/// The segregated witness program is technically only the program bytes _excluding_ the witness
/// version, however we maintain length invariants on the `program` that are governed by the version
/// number, therefore we carry the version number around along with the program bytes.
```